Sources/SwiftRefactor/RefactoringProvider.swift

```diff
@@ -84,3 +96,80 @@ extension RefactoringProvider where Context == Void {
     return self.refactor(syntax: syntax, in: ())
   }
 }
+
+extension SyntaxRefactoringProvider {
+  /// Provides a default implementation for
+  /// `EditRefactoringProvider.textRefactor(syntax:in:)` that produces an edit
+  /// to replace the input of `refactor(syntax:in:)` with its returned output.
+  public static func textRefactor(syntax: Self.Input, in context: Self.Context) -> [Edit] {
+    guard let output = refactor(syntax: syntax, in: context) else {
+      return []
+    }
+    return [Edit.replace(syntax, with: output.description)]
+  }
+}
+
+/// An textual edit to the original source represented by a range and a
+/// replacement.
+public struct Edit: Equatable {
+  /// The location in the original source that this edit starts at.
+  public let start: AbsolutePosition
+  /// The (closed) end of this edit in the original source. Equal to the start
+  /// for an addition.
+  public let end: AbsolutePosition
+  /// The text to replace the original range with. Empty for a deletion.
+  public let replacement: String
+
+  /// Length of the original source range that this edit applies to. Zero if
+  /// this is an addition.
+  public var length: SourceLength {
+    return SourceLength(utf8Length: end.utf8Offset - start.utf8Offset)
+  }
+
+  /// Create an edit to replace `start` to (closed) `end` in the original
+  /// source with `replacement`.
+  public init(start: AbsolutePosition, end: AbsolutePosition, replacement: String) {
+    self.start = start
+    self.end = end
+    self.replacement = replacement
+  }
+
+  /// Convenience function to create a textual addition after the given node
+  /// and its trivia.
+  public static func insert<T: SyntaxProtocol>(_ newText: String, after node: T) -> Edit {
+    return Edit(start: node.endPosition, end: node.endPosition, replacement: newText)
+  }
+
+  /// Convenience function to create a textual addition before the given node
+  /// and its trivia.
+  public static func insert<T: SyntaxProtocol>(_ newText: String, before node: T) -> Edit {
+    return Edit(start: node.position, end: node.position, replacement: newText)
+  }
+
+  /// Convenience function to create a textual replacement of the given node,
+  /// including its trivia.
+  public static func replace<T: SyntaxProtocol>(_ node: T, with replacement: String) -> Edit {
+    return Edit(start: node.position, end: node.endPosition, replacement: replacement)
+  }
+
+  /// Convenience function to create a textual deletion the given node and its
+  /// trivia.
+  public static func remove<T: SyntaxProtocol>(_ node: T) -> Edit {
+    return Edit(start: node.position, end: node.endPosition, replacement: "")
+  }
+}
+
+extension Edit: CustomStringConvertible {
+  public var description: String {
```

Comment:
Seems useful as a description in general to me. Any reason to prefer debug here?

Reply:
I usually try to use `description` only if there is a single, obvious string representation of the type because it’s fairly easy to use `description` in code. And then `description` also needs to provide some sort of stability guarantees IMO. `debugDescription` just makes a lot fewer guarantees IMO and it’s sufficient here.

Reply:
Fair enough.